### PR TITLE
Api implementation

### PR DESCRIPTION
**Pull request by gabriel-v**
Adds API endpoints for https://github.com/liquidinvestigations/liquidinvestigations/issues/114.

This is full of `# TODO`s with what to call next.

### CODE REVIEW

File: liquidcore/config/views.py [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+        serializer = NetworkDomainSerializer(data=request.data)
+        if serializer.is_valid():
+            # TODO send change domain command to system
+            return Response({"detail": "not implemented"},
+                            status=status.HTTP_501_NOT_IMPLEMENTED)
+        return Response(serializer.errors, status=status.HTTP_400_BAD_REQUEST)
+
+class NetworkSettingAPIView(APIView):
+    def get(self, request, format=None):
+        setting = get_object_or_404(Setting, name=self.setting_name)
```
Comment (mgax) [inline liquidcore/config/views.py]: Can we return a blank value instead of 404? Maybe use `get_or_create` to create a default value.

**Comment by mgax** [comment]
`/api/services/` -> the `url` should have `http://` or `https://` in front. Django's request is happy to tell you which one to use.

**Comment by mgax** [comment]
@gabriel-v the UI guys pointed out a couple of bugs:

* 404 error: Not Found: `/api/network/lan/`
* 500 error: NameError: name 'UserSerializer' is not defined

File: liquidcore/config/models.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+class Service(models.Model):
+    name = models.CharField(max_length=30, primary_key=True)
+    is_enabled = models.BooleanField()
+    state = models.CharField(max_length=30, blank=True)
+    state_description = models.TextField(blank=True)
+    error_message = models.TextField(blank=True)
+
+    @property
+    def url(self):
+        return SERVICE_URLS.get(self.name)
```
Comment (mgax) [inline liquidcore/config/models.py:23]: Nice.

File: liquidcore/config/serializers.py [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+from rest_framework import serializers
+
+from .models import *
+
+def valid_username(username):
+    if not username:
+        raise serializers.ValidationError("empty username")
+    if not bool(re.match(r"^(\w|\.)+$", username)):
+        raise serializers.ValidationError("username can only contain letters, digits, underscore and dot")
+    if len(username) > 64:
```
Comment (mgax) [inline liquidcore/config/serializers.py]: Do we have a minimum username length restriction? I think Hypothesis requires at least 2 letters.

File: liquidcore/config/serializers.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+        raise serializers.ValidationError("username too long")
+    return username
+
+def valid_password(password):
+    if len(password) < 8:
+        raise serializers.ValidationError("password too short")
+    return password
+
+def valid_domain(domain):
+    if not bool(re.match(r"^((\w|-)+\.)+(\w+)$", domain)):
```
Comment (mgax) [inline liquidcore/config/serializers.py]: I think `_` is a valid character too. There are probably others, that we don't care about, but `_` might actually be used.

**Review by mgax — APPROVED** [review]
I don't see anything really weird, but honestly there's too much code to grok. Just merge it and we'll see if it breaks. And create an issue to write tests for this API, it's easily testable and very easy to mess something up (authorization, data formats, whatever).

File: liquidcore/config/serializers.py (hunk 4 above)
Comment (gabriel-v) [reply liquidcore/config/serializers.py]: `\w` is alphanumeric + underscore. The "alpha" part permits unicode word characters, so I think we want to limit it to the ASCII `[a-zA-Z0-9_]` equivalent for now.  See https://docs.python.org/3.6/library/re.html for details

File: liquidcore/config/serializers.py (hunk 3 above)
Comment (gabriel-v) [reply liquidcore/config/serializers.py]: I say we fix it at 3 characters just to be safe.

Comment (mgax) [reply liquidcore/config/serializers.py]: Sold, to the man with three fingers!

File: liquidcore/config/serializers.py (hunk 4 above)
Comment (mgax) [reply liquidcore/config/serializers.py]: Yes.

File: liquidcore/config/serializers.py [hunk 5]
```diff
@@ -0,0 +1,127 @@
+import re
+from django.contrib.auth.models import User
+from rest_framework import serializers
+
+from .models import *
+
+USERNAME_REGEX = r"^([a-zA-Z0-9_]|\.)+$"
+USERNAME_URL_REGEX = USERNAME_REGEX[1:-1]
```
Comment (mgax) [inline liquidcore/config/serializers.py:8]: This is weird. If you insist on deduplicating the regexp, do it the other way around:

```python
USERNAME_URL_REGEX = r'([a-zA-Z0-9_]|\.)+'
USERNAME_REGEX = r"^{}$".format(USERNAME_URL_REGEX)
```

File: liquidcore/config/serializers.py [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+    username = serializers.CharField(validators=[valid_username])
+
+    # hyperlink urls using username as identity
+    url = serializers.HyperlinkedIdentityField(
+        view_name='user-detail',
+        lookup_field='username'
+    )
+    class Meta:
+        model = User
+        fields = ('url', 'username', 'first_name',
```
Comment (mgax) [inline liquidcore/config/serializers.py:43]: We should have a test that proves that a user, or an admin, can't change any username.

File: liquidcore/config/serializers.py [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+                  'last_name', 'is_admin', 'is_active')
+
+class UpdateUserSerializer(serializers.HyperlinkedModelSerializer):
```
Comment (mgax) [inline liquidcore/config/serializers.py:46]: So `UpdateUserSerializer` is basically like `CreateUserSerializer` but with a read-only username. Does subclassing work, to remove the duplication?

File: liquidcore/config/serializers.py [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+    # old_password is not validated, as it's just being passed to check_login
+    old_password = serializers.CharField(required=False)
+    new_password = serializers.CharField(validators=[valid_password])
+
+class ServiceSerializer(serializers.ModelSerializer):
+    url = serializers.ReadOnlyField()
+
+    class Meta:
+        model = Service
+        fields = '__all__'
```
Comment (mgax) [inline liquidcore/config/serializers.py:74]: Is this not the default?

File: liquidcore/config/serializers.py [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+
+class ServiceEnabledSerializer(serializers.Serializer):
+    is_enabled = serializers.BooleanField()
+
+class NodeSerializer(serializers.ModelSerializer):
+    data = serializers.ReadOnlyField()
+
+    class Meta:
+        model = Node
+        exclude = ('data_text',)
```
Comment (mgax) [inline liquidcore/config/serializers.py]: I'd prefer we write that as `exclude = ['data_text']`.

File: liquidcore/config/urls.py [hunk 10]
```diff
@@ -0,0 +1,21 @@
+from django.conf.urls import url, include
+from rest_framework import routers
+from rest_framework.urlpatterns import format_suffix_patterns
+
+from .views import *
```
Comment (mgax) [inline liquidcore/config/urls.py:5]: Please don't `import *`. Maybe just `from . import views` and prepend the views with `views.`.

File: liquidcore/config/views.py [hunk 11]
```diff
@@ -0,0 +1,239 @@
...
+from django.conf import settings
+from django.contrib.auth.models import User
+from rest_framework.response import Response
+from rest_framework.permissions import IsAuthenticated, AllowAny, IsAdminUser
+from rest_framework import viewsets
+from rest_framework.decorators import detail_route, list_route, api_view
+from rest_framework.views import APIView
+from rest_framework import status
+
+from .models import *
```
Comment (mgax) [inline liquidcore/config/views.py:12]: Please don't `import *`. Maybe just `from . import models` and prepend the model classes with `models.`.

File: liquidcore/config/views.py [hunk 12]
```diff
@@ -0,0 +1,239 @@
...
+    lookup_value_regex = USERNAME_URL_REGEX
+
+    # don't let the user update the `username` field
+    def get_serializer_class(self):
+        if self.request.method == 'POST':
+            return CreateUserSerializer
+        return UpdateUserSerializer
+
+    # @list_route creates an endpoint that doesn't contain the pk.
+    # We don't return a list, but that's not a problem.
```
Comment (mgax) [inline liquidcore/config/views.py:28]: I utterly fail to understand what you're saying here.

File: liquidcore/config/views.py [hunk 13]
```diff
@@ -0,0 +1,239 @@
...
+    )
+    def whoami(sef, request):
+        """Gets the current logged in user.
+        Marks the result under the key 'is_authenticated'."""
+
+        username = request.user.username
+        if not username:
+            return Response({"is_authenticated": False})
+        queryset = User.objects.get(username=username)
+        data = UpdateUserSerializer(queryset, context={'request':request}).data
```
Comment (mgax) [inline liquidcore/config/views.py:42]: Can't you just do `data = UpdateUserSerializer(request.user, context={'request': request}).data`?

File: liquidcore/config/views.py [hunk 14]
```diff
@@ -0,0 +1,239 @@
...
+    )
+    def active(self, request, username=None):
+        """Sets the user as active or inactive"""
+        user = self.get_object()
+
+        serializer = UserActiveSerializer(data=request.data)
+        serializer.is_valid(raise_exception=True)
+        data = serializer.validated_data
+
+        user.is_active = data['is_active']
```
Comment (mgax) [inline liquidcore/config/views.py:58]: Do we have a view to update `is_admin`?

File: testsuite/test_config_users.py [hunk 15]
```diff
@@ -0,0 +1,209 @@
...
+
+    # change the password
+    password = client.post("/api/users/admin/password/", data={
+        "new_password": "12345678",
+    })
+    assert password.status_code == 200
+
+    # make sure new login works
+    client.logout()
+    client.login(username='admin', password='12345678')
```
Comment (mgax) [inline testsuite/test_config_users.py]: I think you mean `assert client.login()`, otherwise the call is pointless.

File: liquidcore/home/urls.py [hunk 16]
```diff
@@ -2,5 +2,5 @@
 from . import views
 
 urlpatterns = [
-    url(r'^$', views.homepage),
+    url(r'^core$', views.homepage),
```
Comment (mgax) [inline liquidcore/home/urls.py]: Let's undo this change.

File: liquidcore/config/system.py [hunk 17]
```diff
@@ -0,0 +1,25 @@
...
+        'wan': settings['wan'],
+        'ssh': settings['ssh'],
+        'services': {
+            s.name: {'enabled': s.is_enabled}
+            for s in models.Service.objects.all()
+        },
+    }
+
+
+def update_system():
```
Comment (mgax) [inline liquidcore/config/system.py:23]: Let's call this `configure_system()`.

File: liquidcore/site/urls.py [hunk 18]
```diff
@@ -7,5 +7,7 @@
     url(r'^accounts/', include('django.contrib.auth.urls')),
     url(r'^accounts/profile$', views.profile),
     url(r'^o/', include('oauth2_provider.urls', namespace='oauth2_provider')),
+    url(r'^api-auth/', include('rest_framework.urls', namespace='rest_framework')),
```
Comment (mgax) [inline liquidcore/site/urls.py:10]: Do we need this mapping? We're using the regular login (`/accounts`) already.

File: testsuite/test_config_network.py [hunk 19]
```diff
@@ -0,0 +1,223 @@
...
+
+@pytest.mark.parametrize("key,vals,broken_vals", [
+    ("lan", [LAN1, LAN2], [LAN_BROKEN]),
+    ("wan", [WAN1, WAN2], [WAN_BROKEN]),
+    ("ssh", [SSH1, SSH2], [SSH_BROKEN]),
+    ("domain", [{'domain': d} for d in GOOD_DOMAINS],
+               [{'domain': d} for d in BAD_DOMAINS]),
+])
+def test_network_configuration(client, key, vals, broken_vals):
+    endpoint = '/api/network/{}/'.format(key)
```
Comment (mgax) [inline testsuite/test_config_network.py]: This is so weird. Why don't you just pass in the whole endpoint url? It's not so long.

File: testsuite/test_config_network.py [hunk 20]
```diff
@@ -0,0 +1,223 @@
...
+    'other~things.xxx',
+    'dash-where-it.shouldnt-be',
+    'put+both-patches.cn',
+]
+
+@pytest.mark.parametrize("key,vals,broken_vals", [
+    ("lan", [LAN1, LAN2], [LAN_BROKEN]),
+    ("wan", [WAN1, WAN2], [WAN_BROKEN]),
+    ("ssh", [SSH1, SSH2], [SSH_BROKEN]),
+    ("domain", [{'domain': d} for d in GOOD_DOMAINS],
```
Comment (mgax) [inline testsuite/test_config_network.py]: How about changing `GOOD_DOMAINS` to this, so the fixture is shorter and fits on a single line?

```python
GOOD_DOMAINS = [{'domain': d} for d in [
    'liquidnode.local',
    'however-bored-ill-write-tests.com',
    'more.segments.please.org',
    'some-dashes.auto',
    'weird-tld.liquid',
    'num8berx.zero',
]]
```

File: testsuite/test_config_users.py [hunk 21]
```diff
@@ -0,0 +1,209 @@
...
+
+def test_password_change_nonadmin(client):
```
Comment (mgax) [inline testsuite/test_config_users.py]: Ok, this test is super useful, but do we have a test for when a user tries to change somebody else's password?

**Review by mgax — CHANGES_REQUESTED** [review]
Basically there should be a test for each view, and for each `if` inside each view.

File: liquidcore/config/serializers.py (hunk 8 above)
Comment (gabriel-v) [reply liquidcore/config/serializers.py:74]: no

File: liquidcore/config/serializers.py (hunk 7 above)
Comment (gabriel-v) [reply liquidcore/config/serializers.py:46]: Yeah, totally forgot about that, thanks

**Comment by mgax** [comment]
@gabriel-v please add issues for whatever was left in the PR review.
